packages/data-sdk/src/Device.ts

```diff
@@ -251,74 +246,81 @@ export class Device extends EventEmitter implements IRealtimeDevice {
+
+    if ("isReady" in rtcClient) {
+      while (!rtcClient.isReady()) {
+        await delay(100);
       }
+    }
 
-      if ("isReady" in rtcClient) {
-        while (!rtcClient.isReady()) {
-          await delay(100);
-        }
+    // WebRTC requires a signaling phase when forming a new connection.
+
+    this.remoteDevicePeerId = await this.getRemoteDevicePeerId(rtcClient);
+
+    const sessionId = await this.createSession(rtcClient);
+
+    // Wait for the signaling process to complete...
+    if (!sessionId) {
+      try {
+        // cleanup
+        this.remoteDevicePeerId = null;
+        await rtcClient.shutdown();
+      } catch (err) {
+        console.error("rtcClient cannot shutdown", { err });
       }
 
-      // WebRTC requires a signaling phase when forming a new connection.
-
-      this.remoteDevicePeerId = await this.getRemoteDevicePeerId(rtcClient);
-
-      const sessionId = await this.createSession(rtcClient);
-
-      // Wait for the signaling process to complete...
-      if (!!sessionId) {
-        const retries = 100;
-        for (let i = 0; i < retries; i++) {
-          const connectionCompleted =
-            rtcClient.getConnectionStatus(this.remoteDevicePeerId) ===
-            "connected";
-          if (connectionCompleted) {
-            console.debug(
-              `${new Date().toISOString()} :: Connection completed after ${i} retries`
-            );
-            this.initConnectionMonitoring();
-            this.rtcClient = rtcClient;
-            this.emit("connect");
-            return;
-          }
-          await delay(100);
-        }
-        throw new Error(
-          "A session was created, but the connection could not be established, possibly due to network issues or misconfigured settings."
-        );
-      } else {
-        throw new Error(`Unable to establish a connection at this time.`);
+      throw new Error(`Unable to establish a connection at this time.`);
+    }
+
+    const retries = 100;
+    for (let i = 0; i < retries; i++) {
+      const connectionCompleted =
+        rtcClient.getConnectionStatus(this.remoteDevicePeerId) === "connected";
+
+      if (!connectionCompleted) {
+        await delay(100);
+        continue;
       }
-    } else {
-      throw new Error(
-        `Already created realtime connection to device ${this.id}`
+
+      console.debug(
+        `${new Date().toISOString()} :: Connection completed after ${i} retries`
       );
+      this.initConnectionMonitoring();
+      this.rtcClient = rtcClient;
+      this.emit("connect");
+      return;
+    }
+
+    console.error(
+      "Connection failed: A session was created, but the connection could not be established, " +
+        "possibly due to network issues or misconfigured settings."
+    );
+
+    try {
+      // cleanup
+      this.remoteDevicePeerId = null;
+      await rtcClient.shutdown();
+    } catch (err) {
+      console.error("rtcClient cannot shutdown", { err });
     }
```

Comment:
I'm not _entirely_ sure what should happen here, given that we've exhausted the connection limit here. Especially in greenfields case they kind of _assume_ that its going to work/establish a connection, just a _log_ that it didn't work might be "less than ideal"